### ISSUE TITLE
From where did self.anchors in the _make_grid method of the Detect module come from?

### ISSUE THREAD

**Issue by edargham** [ultralytics/yolov5#9683]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I'm reading through the yolo.py code, trying to make sense of everything that was done. I stopped at the `_make_grid(...)` method of the `Detect(nn.Module)` class and couldn't help but notice the `self.anchors[i]` being used is never mentioned anywhere else in the class. 

Is this the same `'anchors'` argument that we registered into the buffer in the class constructor? If not where is this element defined?

Code Reference in question:

```python
class Detect(nn.Module):
    stride = None  # strides computed during build
    dynamic = False  # force grid reconstruction
    export = False  # export mode

    def __init__(self, nc=80, anchors=(), ch=(), inplace=True):  # detection layer
        super().__init__()
        self.nc = nc  # number of classes
        self.no = nc + 5  # number of outputs per anchor
        self.nl = len(anchors)  # number of detection layers
        self.na = len(anchors[0]) // 2  # number of anchors
        self.grid = [torch.empty(1)] * self.nl  # init grid
        self.anchor_grid = [torch.empty(1)] * self.nl  # init anchor grid

        #######################################################################
        # ONLY OTHER PLACE WHERE anchors IS DEFINED / USED.
        self.register_buffer('anchors', torch.tensor(anchors).float().view(self.nl, -1, 2))  # shape(nl,na,2)
        #######################################################################

        self.m = nn.ModuleList(nn.Conv2d(x, self.no * self.na, 1) for x in ch)  # output conv
        self.inplace = inplace  # use inplace ops (e.g. slice assignment)
    .
    .
    .
    def _make_grid(self, nx=20, ny=20, i=0, torch_1_10=check_version(torch.__version__, '1.10.0')):
        ## ################################################
        # USED WITHOUT BEING DEFINED ANYWHERE ELSE IN THE CLASS.
        d = self.anchors[i].device
        t = self.anchors[i].dtype
        ###################################################

        shape = 1, self.na, ny, nx, 2  # grid shape
        y, x = torch.arange(ny, device=d, dtype=t), torch.arange(nx, device=d, dtype=t)
        yv, xv = torch.meshgrid(y, x, indexing='ij') if torch_1_10 else torch.meshgrid(y, x)  # torch>=0.7 compatibility
        grid = torch.stack((xv, yv), 2).expand(shape) - 0.5  # add grid offset, i.e. y = 2.0 * x - 0.5
        anchor_grid = (self.anchors[i] * self.stride[i]).view((1, self.na, 1, 1, 2)).expand(shape)
        return grid, anchor_grid
    .
    .
    .
```

**Comment by glenn-jocher** [comment]
@edargham yes anchors are passed to init function and registered as a buffer before being used in _make_grid()

**Comment by edargham** [comment]
I see that makes sense.
What’s the advantage in registering the anchors into the buffer over storing them as a class property for the Detect module in `self.anchors`?

**Comment by glenn-jocher** [comment]
buffers and parameters move to different devices with model.to()

**Comment by edargham** [comment]
I see that makes sense. By calling `self.register_buffer(…)` it’s delegating the storage of the arguments to the underlying torch backend library if I’m not mistaken, kindly correct me if I’m wrong. I’m still green with PyTorch so sorry if that sounds trivial.

**Comment by glenn-jocher** [comment]
@edargham you're absolutely right! By using `self.register_buffer(…)`, we delegate the storage of the anchors to PyTorch's backend, allowing essential functionality like moving the model to different devices using `model.to()`. Your understanding is spot on! And there's no need to apologize, we all start somewhere. Keep up the great work!
